substrate/serai/client/src/lib.rs

```diff
@@ -0,0 +1,77 @@
+use thiserror::Error;
+
+use serde::Serialize;
+use scale::Decode;
+
+use subxt::{tx::BaseExtrinsicParams, Config as SubxtConfig, OnlineClient};
+
+pub use serai_primitives as primitives;
+use primitives::{Signature, NativeAddress};
+
+use serai_runtime::{system::Config, Runtime};
+
+pub mod in_instructions;
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub(crate) struct SeraiConfig;
+impl SubxtConfig for SeraiConfig {
+  type BlockNumber = <Runtime as Config>::BlockNumber;
```

Comment:
Why not just use the block struct? for these and below.

Reply:
The `BlockNumber(u32)` struct? Because Substrate requires a lot of traits on these types. I tried implementing it for `Amount` and it was ~300 lines of bs which broke several of the goals of the `Amount` type (float support, some math ops you likely shouldn't be using with `Amounts` is an uber-traditional system...).

Reply:
#[derive(Debug, Clone)]
pub struct Block {
	pub hash: Hash,
	pub parent_hash: Hash,
	pub number: BlockNumber,
}

Reply:
We don't have such a struct right now. My reason against it is IDs are optimally Copy-able. That exceeds intended Copy bounds. It's also shotgunning data and suggests it's fine to use the number, when it frequently isn't.

Reply:
To be clear, the processor is currently coded for numbers. I wouldn't mind moving it to block hashes. Since it's only operating past the confirmation depth it... *should* be fine? to use numbers, if we want to save the dev effort.

I'd also note that the recent Substrate upgrades have been removing usage *and* usability of numbers as IDs. I only believe in their usage for finalized data, and then as a project, we assume things sufficiently confirmed are final. Shotgunning them provides their accessibility where they arguably shouldn't be (and one of the ways Substrate removed their usability was by removing a variant of hash OR number).

I'd also note I did rewrite the Monero RPC to be hash-based when I wrote the Scanner scratch: https://github.com/serai-dex/serai/commit/1d6df0099ce62a5f80c00791668e8b7a46e43ffd

Reply:
We should use hashes.